crates/rust-analyzer/src/global_state.rs

```diff
@@ -350,10 +377,30 @@ impl GlobalState {
             }
             let (vfs, line_endings_map) = &mut *RwLockUpgradableReadGuard::upgrade(guard);
             bytes.into_iter().for_each(|(file_id, text)| match text {
-                None => change.change_file(file_id, None),
+                None => {
+                    change.change_file(file_id, None);
+                    if let Some(vfs_path) = modified_ratoml_files.get(&file_id) {
+                        if vfs_path == self.config.user_config_path() {
+                            user_config_file = Some((file_id, None));
+                        } else if vfs_path == &root_vfs_path {
+                            root_path_ratoml = Some((file_id, None));
+                        } else {
+                            ratoml_text_map.insert(file_id, (vfs_path.clone(), None));
+                        }
+                    }
+                }
                 Some((text, line_endings)) => {
                     line_endings_map.insert(file_id, line_endings);
-                    change.change_file(file_id, Some(text));
+                    change.change_file(file_id, Some(text.clone()));
+                    if let Some(vfs_path) = modified_ratoml_files.get(&file_id) {
+                        if vfs_path == self.config.user_config_path() {
+                            user_config_file = Some((file_id, Some(text.clone())));
+                        } else if vfs_path == &root_vfs_path {
+                            root_path_ratoml = Some((file_id, Some(text.clone())));
+                        } else {
+                            ratoml_text_map.insert(file_id, (vfs_path.clone(), Some(text.clone())));
+                        }
+                    }
```

Comment:
Can be simplified by having the lbocks return `None` or `Some(text)`, assign that to a  var and then branching afterwards using the var output. The only difference in the two arms is the insertion to `line_endings_map` after all